lib/tree.rb

```diff
@@ -16,49 +16,153 @@ def initialize
+    else
+      find_helper(current_node.right, search)
+    end
+  end
+  
   def find(key)
-    raise NotImplementedError
+    found = find_helper(@root, key)
+    return found.value if found
+  end
+
+  # Time Complexity: O(n) where n is number of nodes in tree
+  # Space Complexity: O(n) where n is number of nodes in tree
+  # return array of all nodes in order
+  # when? if you need to print all the nodes of a tree in order 
+  def inorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    inorder_helper(current_node.left, list) if current_node.left
+    list << {key: current_node.key, value: current_node.value}
+    inorder_helper(current_node.right, list) if current_node.right
+    
+    return list
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) where n is number of nodes in tree
+  # Space Complexity: O(n) where n is number of nodes in tree
+  # when? if you need to save a tree data structure to disk, or send it across the network and maintain the structure
+  def preorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    list << {key: current_node.key, value: current_node.value}
+    preorder_helper(current_node.left, list) if current_node.left
+    preorder_helper(current_node.right, list) if current_node.right
+    
+    return list
+  end
+  
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
+  end
+
+  # Time Complexity: O(n) where n is number of nodes in tree
+  # Space Complexity: O(n) where n is number of nodes in tree
+  # when? if you need to delete all the nodes in BST
+  def postorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    postorder_helper(current_node.left, list) if current_node.left
+    postorder_helper(current_node.right, list) if current_node.right
+    list << {key: current_node.key, value: current_node.value}
+    
+    return list
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) where n is number of nodes in tree
+  # Space Complexity: O(n) where n is number of nodes in tree
+  def height_helper(current_node, count, max)
+    return max if current_node.nil?
+
+    max = count if count > max
+    height_helper(current_node.left, count + 1, max)
+    height_helper(current_node.right, count + 1, max)
+  end
+  
   def height
-    raise NotImplementedError
+    return height_helper(@root, 1, 0)
   end
 
   # Optional Method
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n) where n is number of nodes in tree
+  # Space Complexity: O(n) where n is number of nodes in tree
+  def bfs_helper(node, list, level)
```

Comment:
This isn't a traditional breadth-first search but it is an interesting way to get the tree level-by-level.